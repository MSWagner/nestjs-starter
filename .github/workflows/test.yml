name: Test

on: [push, pull_request]

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose-test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs ananas-dark-backend-postgres-test

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '10.13.0'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test