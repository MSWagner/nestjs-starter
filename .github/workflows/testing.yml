name: Testing

on: [push, pull_request]

jobs:
  test-local:
    name: Testing Local
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose.test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test-standalone

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test

  test-docker:
    name: Testing Docker Container
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      
      - name: Create docker network
        run: docker network create project-name-network-dev
      - name: Start Docker Dev Container
        run: yarn docker:up:hidden
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test

      - name: Run database migrations
        run: yarn docker:db:migrate

      - name: Start tests in docker container
        run: yarn docker:test