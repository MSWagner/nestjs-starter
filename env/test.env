NODE_ENV=test

PGHOST=localhost
PGPORT=5431
PGUSER=root
PGPASSWORD=toor
PGDATABASE=project-name-test-db