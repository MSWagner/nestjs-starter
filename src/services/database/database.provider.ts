import { Connection, createConnection, getConnection, getConnectionManager } from "typeorm";

import CONFIG from "../../config";

export const DatabaseProvider = {
    provide: CONFIG.database.defaultConnectionName,
    useFactory: async (): Promise<Connection> => {
        try {
            return getConnection();
        } catch (error) {
            // If AlreadyHasActiveConnectionError occurs, return already existent connection
            if (error.name === "AlreadyHasActiveConnectionError") {
                const existentConn = getConnectionManager().get("default");
                return existentConn;
            }

            return createConnection({
                type: "postgres",
                host: process.env.PGHOST,
                port: parseInt(process.env.PGPORT, 10),
                username: process.env.PGUSER,
                password: process.env.PGPASSWORD,
                database: process.env.PGDATABASE,
                entities: [__dirname + "/../../**/*.entity.{ts,js}"],
                migrations: [__dirname + "/../../migrations/*.ts"],
                cli: { migrationsDir: `${__dirname + "/../../migrations"}` },
                synchronize: false,
                logging: CONFIG.env.isDevelopment ? ["info"] : false
            });
        }
    }
};
