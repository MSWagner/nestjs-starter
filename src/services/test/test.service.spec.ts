import { Test, TestingModule } from "@nestjs/testing";
import { TestService } from "./test.service";
import { DatabaseModule } from "../database/database.module";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import { fixtureTrees } from "./fixtures";

describe("TestService", () => {
    let service: TestService;

    beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [TestService, DatabaseModule]
        }).compile();

        service = module.get<TestService>(TestService);
    });

    afterAll((done) => {
        service.connectionManager.connection.close();
        done();
    });

    it("should check the orderIds of the User, AccessToken & RefreshToken entities", () => {
        const userOrder = service.getOrder("User");
        const refreshTokenOrder = service.getOrder("RefreshToken");
        const accessTokenOrder = service.getOrder("AccessToken");

        expect(userOrder).toEqual(1);
        expect(refreshTokenOrder).toEqual(3);
        expect(accessTokenOrder).toEqual(4);
    });

    it("should return all entities", async () => {
        const entities = await service.getEntities();

        expect(entities).toMatchSnapshot("DatabaseEntitiyList");
    });

    it("should reload the fixtures without a additional user from before", async () => {
        const newUser = new User();
        newUser.username = "TestUser";
        await newUser.save();

        const newUserCountBefore = await User.count({ username: "TestUser" });
        expect(newUserCountBefore).toEqual(1);

        await service.reloadFixtures();

        const newUserCountAfter = await User.count({ username: "TestUser" });
        expect(newUserCountAfter).toEqual(0);
    });

    it("should remove all fixtures & load them again into the db", async () => {
        const userCountBefore = await User.count();
        expect(userCountBefore).toEqual(fixtureTrees.User.length);

        const refreshTokenCountBefore = await RefreshToken.count();
        expect(refreshTokenCountBefore).toEqual(fixtureTrees.RefreshToken.length);

        const accessTokenCountBefore = await AccessToken.count();
        expect(accessTokenCountBefore).toEqual(fixtureTrees.AccessToken.length);

        const entities = await service.getEntities();
        await service.cleanAll(entities);

        const userCountAfterClean = await User.count();
        expect(userCountAfterClean).toEqual(0);

        const refreshTokenCountAfterClean = await RefreshToken.count();
        expect(refreshTokenCountAfterClean).toEqual(0);

        const accessTokenCountAfterClean = await AccessToken.count();
        expect(accessTokenCountAfterClean).toEqual(0);

        await service.loadAll(entities);

        const userCountAfterLoad = await User.count();
        expect(userCountAfterLoad).toEqual(fixtureTrees.User.length);

        const refreshTokenCountAfterLoad = await RefreshToken.count();
        expect(refreshTokenCountAfterLoad).toEqual(fixtureTrees.RefreshToken.length);

        const accessTokenCountAfterLoad = await AccessToken.count();
        expect(accessTokenCountAfterLoad).toEqual(fixtureTrees.AccessToken.length);
    });

    it("should replace all uuids in the object", async () => {
        const object = [
            {
                test: {
                    i1: "d3fd0451-ec15-4f60-8e64-9142c4ab3a89"
                },
                test2: [[{ test3: "020bf68b-23bd-42b6-92b9-c6e0e897eb8f" }, "757d21f0-f72c-4dc4-a6d4-badd1305ea7a"]]
            },
            "44218412-bfe7-4c70-aeca-276c0fcf099e",
            {
                test: {
                    i1: "2019-09-28 15:26:33.186"
                },
                test2: [[{ test3: "2019-09-28 15:26:33.186" }, "2019-09-28 15:26:33.186"]]
            },
            "2019-09-28 15:26:33.186"
        ];

        const replacedObject = service.replaceUUIDs(object);

        expect(replacedObject).toMatchSnapshot("UUIDsReplaced");
    });

    it("should replace all dates in the object", async () => {
        const object = [
            {
                test: {
                    i1: "d3fd0451-ec15-4f60-8e64-9142c4ab3a89"
                },
                test2: [[{ test3: "020bf68b-23bd-42b6-92b9-c6e0e897eb8f" }, "757d21f0-f72c-4dc4-a6d4-badd1305ea7a"]]
            },
            "44218412-bfe7-4c70-aeca-276c0fcf099e",
            {
                test: {
                    i1: "2019-09-28 15:26:33.186"
                },
                test2: [[{ test3: "2019-09-28 15:26:33.186" }, "2019-09-28 15:26:33.186"]]
            },
            "2019-09-28 15:26:33.186"
        ];

        const replacedObject = service.replaceDates(object);

        expect(replacedObject).toMatchSnapshot("DateReplaced");
    });

    it("should replace all dates & uuids in the object", async () => {
        const object = [
            {
                test: {
                    i1: "d3fd0451-ec15-4f60-8e64-9142c4ab3a89"
                },
                test2: [[{ test3: "020bf68b-23bd-42b6-92b9-c6e0e897eb8f" }, "757d21f0-f72c-4dc4-a6d4-badd1305ea7a"]]
            },
            "44218412-bfe7-4c70-aeca-276c0fcf099e",
            {
                test: {
                    i1: "2019-09-28 15:26:33.186"
                },
                test2: [[{ test3: "2019-09-28 15:26:33.186" }, "2019-09-28 15:26:33.186"]]
            },
            "2019-09-28 15:26:33.186"
        ];

        const replacedObject = service.replaceValues(object, ["DATE", "UUID"]);

        expect(replacedObject).toMatchSnapshot("ReplaceDateAndUUIDs");
    });
});
