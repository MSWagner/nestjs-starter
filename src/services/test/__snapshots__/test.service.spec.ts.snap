// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestService should replace all dates & uuids in the object: ReplaceDateAndUUIDs 1`] = `
[
  {
    "test": {
      "i1": "i1",
    },
    "test2": [
      [
        {
          "test3": "test3",
        },
        "UUID",
      ],
    ],
  },
  "UUID",
  {
    "test": {
      "i1": "i1",
    },
    "test2": [
      [
        {
          "test3": "test3",
        },
        "DATE",
      ],
    ],
  },
  "DATE",
]
`;

exports[`TestService should replace all dates in the object: DateReplaced 1`] = `
[
  {
    "test": {
      "i1": "d3fd0451-ec15-4f60-8e64-9142c4ab3a89",
    },
    "test2": [
      [
        {
          "test3": "020bf68b-23bd-42b6-92b9-c6e0e897eb8f",
        },
        "757d21f0-f72c-4dc4-a6d4-badd1305ea7a",
      ],
    ],
  },
  "44218412-bfe7-4c70-aeca-276c0fcf099e",
  {
    "test": {
      "i1": "i1",
    },
    "test2": [
      [
        {
          "test3": "test3",
        },
        "DATE",
      ],
    ],
  },
  "DATE",
]
`;

exports[`TestService should replace all uuids in the object: UUIDsReplaced 1`] = `
[
  {
    "test": {
      "i1": "i1",
    },
    "test2": [
      [
        {
          "test3": "test3",
        },
        "UUID",
      ],
    ],
  },
  "UUID",
  {
    "test": {
      "i1": "2019-09-28 15:26:33.186",
    },
    "test2": [
      [
        {
          "test3": "2019-09-28 15:26:33.186",
        },
        "2019-09-28 15:26:33.186",
      ],
    ],
  },
  "2019-09-28 15:26:33.186",
]
`;

exports[`TestService should return all entities: DatabaseEntitiyList 1`] = `
[
  {
    "name": "RefreshToken",
    "order": 3,
    "tableName": "refresh_token",
  },
  {
    "name": "AccessToken",
    "order": 4,
    "tableName": "access_token",
  },
  {
    "name": "Permission",
    "order": 0,
    "tableName": "permission",
  },
  {
    "name": "UserPermission",
    "order": 2,
    "tableName": "user_permission",
  },
  {
    "name": "PushToken",
    "order": 5,
    "tableName": "push_token",
  },
  {
    "name": "User",
    "order": 1,
    "tableName": "user",
  },
]
`;
